test/integration/controller_instance_test.go

```diff
@@ -992,3 +1055,470 @@ func TestCreateServiceInstanceWithProvisionFailure(t *testing.T) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async:        true,
+						OperationKey: &key,
+					},
+				}
+			},
+		},
+		{
+			name: "asynchronous deprovision with operation key, no binding",
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async:        true,
+						OperationKey: &key,
+					},
+				}
+			},
+		},
+		{
+			name:    "asynchronous deprovision without operation key",
+			binding: getTestBinding(),
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async: true,
+					},
+				}
+			},
+		},
+		{
+			name: "asynchronous deprovision without operation key, no binding",
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{
+						Async: true,
+					},
+				}
+			},
+		},
+		{
+			name:    "deprovision instance with binding not deleted",
+			binding: getTestBinding(),
+			setup: func(ct *controllerTest) {
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{},
+				}
+			},
+			testFunction: func(ct *controllerTest) {
+				if err := ct.client.ServiceInstances(ct.instance.Namespace).Delete(ct.instance.Name, &metav1.DeleteOptions{}); err != nil {
+					ct.t.Fatalf("instance delete should have been accepted: %v", err)
+				}
+
+				condition := v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionReady,
+					Status: v1beta1.ConditionFalse,
+					Reason: "DeprovisionBlockedByExistingCredentials",
+				}
+				if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, condition); err != nil {
+					ct.t.Fatalf("error waiting for instance condition: %v", err)
+				}
+			},
+		},
+		{
+			name: "deprovision instance during in progress provision",
+			skipVerifyingInstanceSuccess: true,
+			setup: func(ct *controllerTest) {
+				ct.osbClient.PollLastOperationReaction = fakeosb.DynamicPollLastOperationReaction(
+					getLastOperationResponseByPollCountStates(2, []osb.LastOperationState{osb.StateInProgress, osb.StateSucceeded}))
+				ct.osbClient.ProvisionReaction = &fakeosb.ProvisionReaction{
+					Response: &osb.ProvisionResponse{
+						Async: true,
+					},
+				}
+				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
+					Response: &osb.DeprovisionResponse{},
+				}
+			},
+			testFunction: func(ct *controllerTest) {
+				verifyCondition := v1beta1.ServiceInstanceCondition{
+					Type:   v1beta1.ServiceInstanceConditionReady,
+					Status: v1beta1.ConditionFalse,
+					Reason: "ProvisionRequestInFlight",
+				}
+				if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, verifyCondition); err != nil {
```

Comment:
This is going to be flaky. We are relying on it taking a few seconds for the controller to go through the two in-progress poll operations. Instead we should not change the polling response to Succeeded until after the test has verified the ProvisionRequestInFlight condition.

Reply:
Ok done